### ISSUE TITLE
Upgrade from Catalina (v1.9x) to Big Sur (v2x) encrypted dataset permanent errors detected even after scrub repair on v1.9

### ISSUE THREAD

**Issue by greg-fischer** [openzfsonosx/openzfs#100]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| MacOS
Distribution Version	| Big Sur (11.5x)
Linux Kernel	|
Architecture	|
ZFS Version	| zfs-macOS-2.1.0-1
SPL Version	|
```
zfs --version
zfs-macOS-2.1.0-1
zfs-kmod-2.1.0-1
```
### Describe the problem you're observing

I recently upgraded from Catalina (on two systems) where I had ZFS 1.9.x to Big Sur and installed the MacOS ZFS 2.1.0 package.  My zpool was not reporting any errors before the upgrade.  It does have a few encrypted datasets.  While attempting import of the pool I receive "Input/output error" on each of the encrypted datasets.  I also noticed "zpool status -v" reports "Permanent errors have been detected in the following files" and lists the 3 encrypted datasets:
```
zebbo/MACDATA-1/ZEBBO-NOBU:<0x0>
zebbo/MACDATA-1/ZEBBO-VM:<0x0>
zebbo/MACDATA-1/ZEBBO:<0x0>
```

It's not listing corrupted files, but the entire dataset.  No checksum errors reports.

Then, I took the pool to another system with Catalina (with ZFS 1.9.2) still installed and scrubbed it.  The errors went away, everything worked normally.  However, when importing again on Big Sur, the same issue came back.  I'm unable to open the encrypted datasets on Big Sur.

Also, I checked on Linux using ZFS 0.8.1 (CentOS 7) and I was able to import, load keys, and mount successfully.  But a scrub did not remove the errors.

### Describe how to reproduce the problem

Use an encrypted dataset starting from v1.9x on Catalina, then take the pool to Big Sur with 2.1.x and see if the encrypted datasets load keys and mount or receive the "permanent errors" notice.

Taking the pool back to Catalina and scrubbing removes the errors, still allows mounts and keys to load, but taking back to Big Sur it fails.

Here's  the ZFS properties on one of the datasets if that helps:
```
zfs get all zebbo/MACDATA-1/ZEBBO
NAME                   PROPERTY                        VALUE                           SOURCE
zebbo/MACDATA-1/ZEBBO  type                            filesystem                      -
zebbo/MACDATA-1/ZEBBO  creation                        Thu Dec  5 19:56 2019           -
zebbo/MACDATA-1/ZEBBO  used                            13.3G                           -
zebbo/MACDATA-1/ZEBBO  available                       612G                            -
zebbo/MACDATA-1/ZEBBO  referenced                      13.3G                           -
zebbo/MACDATA-1/ZEBBO  compressratio                   1.46x                           -
zebbo/MACDATA-1/ZEBBO  mounted                         no                              -
zebbo/MACDATA-1/ZEBBO  quota                           none                            default
zebbo/MACDATA-1/ZEBBO  reservation                     none                            default
zebbo/MACDATA-1/ZEBBO  recordsize                      128K                            default
zebbo/MACDATA-1/ZEBBO  mountpoint                      /Volumes/ZEBBO                  local
zebbo/MACDATA-1/ZEBBO  sharenfs                        off                             default
zebbo/MACDATA-1/ZEBBO  checksum                        edonr                           inherited from zebbo
zebbo/MACDATA-1/ZEBBO  compression                     lz4                             inherited from zebbo
zebbo/MACDATA-1/ZEBBO  atime                           off                             inherited from zebbo
zebbo/MACDATA-1/ZEBBO  devices                         on                              default
zebbo/MACDATA-1/ZEBBO  exec                            on                              default
zebbo/MACDATA-1/ZEBBO  setuid                          on                              default
zebbo/MACDATA-1/ZEBBO  readonly                        off                             default
zebbo/MACDATA-1/ZEBBO  zoned                           off                             default
zebbo/MACDATA-1/ZEBBO  snapdir                         hidden                          default
zebbo/MACDATA-1/ZEBBO  aclmode                         discard                         default
zebbo/MACDATA-1/ZEBBO  aclinherit                      restricted                      default
zebbo/MACDATA-1/ZEBBO  createtxg                       26                              -
zebbo/MACDATA-1/ZEBBO  canmount                        on                              default
zebbo/MACDATA-1/ZEBBO  xattr                           sa                              inherited from zebbo
zebbo/MACDATA-1/ZEBBO  copies                          1                               default
zebbo/MACDATA-1/ZEBBO  version                         5                               -
zebbo/MACDATA-1/ZEBBO  utf8only                        on                              -
zebbo/MACDATA-1/ZEBBO  normalization                   formD                           -
zebbo/MACDATA-1/ZEBBO  casesensitivity                 insensitive                     -
zebbo/MACDATA-1/ZEBBO  vscan                           off                             default
zebbo/MACDATA-1/ZEBBO  nbmand                          off                             default
zebbo/MACDATA-1/ZEBBO  sharesmb                        off                             default
zebbo/MACDATA-1/ZEBBO  refquota                        none                            default
zebbo/MACDATA-1/ZEBBO  refreservation                  none                            default
zebbo/MACDATA-1/ZEBBO  guid                            5802056370195559715             -
zebbo/MACDATA-1/ZEBBO  primarycache                    all                             default
zebbo/MACDATA-1/ZEBBO  secondarycache                  all                             default
zebbo/MACDATA-1/ZEBBO  usedbysnapshots                 428K                            -
zebbo/MACDATA-1/ZEBBO  usedbydataset                   13.3G                           -
zebbo/MACDATA-1/ZEBBO  usedbychildren                  0B                              -
zebbo/MACDATA-1/ZEBBO  usedbyrefreservation            0B                              -
zebbo/MACDATA-1/ZEBBO  logbias                         latency                         default
zebbo/MACDATA-1/ZEBBO  objsetid                        147                             -
zebbo/MACDATA-1/ZEBBO  dedup                           off                             default
zebbo/MACDATA-1/ZEBBO  mlslabel                        none                            default
zebbo/MACDATA-1/ZEBBO  sync                            standard                        default
zebbo/MACDATA-1/ZEBBO  dnodesize                       legacy                          default
zebbo/MACDATA-1/ZEBBO  refcompressratio                1.46x                           -
zebbo/MACDATA-1/ZEBBO  written                         13.3G                           -
zebbo/MACDATA-1/ZEBBO  logicalused                     19.5G                           -
zebbo/MACDATA-1/ZEBBO  logicalreferenced               19.5G                           -
zebbo/MACDATA-1/ZEBBO  volmode                         default                         default
zebbo/MACDATA-1/ZEBBO  filesystem_limit                none                            default
zebbo/MACDATA-1/ZEBBO  snapshot_limit                  none                            default
zebbo/MACDATA-1/ZEBBO  filesystem_count                none                            default
zebbo/MACDATA-1/ZEBBO  snapshot_count                  none                            default
zebbo/MACDATA-1/ZEBBO  snapdev                         hidden                          default
zebbo/MACDATA-1/ZEBBO  acltype                         nfsv4                           default
zebbo/MACDATA-1/ZEBBO  context                         none                            default
zebbo/MACDATA-1/ZEBBO  fscontext                       none                            default
zebbo/MACDATA-1/ZEBBO  defcontext                      none                            default
zebbo/MACDATA-1/ZEBBO  rootcontext                     none                            default
zebbo/MACDATA-1/ZEBBO  relatime                        off                             default
zebbo/MACDATA-1/ZEBBO  redundant_metadata              all                             default
zebbo/MACDATA-1/ZEBBO  overlay                         on                              default
zebbo/MACDATA-1/ZEBBO  encryption                      aes-256-ccm                     -
zebbo/MACDATA-1/ZEBBO  keylocation                     file:///Users/greg/.zfs.key     local
zebbo/MACDATA-1/ZEBBO  keyformat                       passphrase                      -
zebbo/MACDATA-1/ZEBBO  pbkdf2iters                     350000                          -
zebbo/MACDATA-1/ZEBBO  encryptionroot                  zebbo/MACDATA-1/ZEBBO           -
zebbo/MACDATA-1/ZEBBO  keystatus                       available                       -
zebbo/MACDATA-1/ZEBBO  special_small_blocks            0                               default
zebbo/MACDATA-1/ZEBBO  com.apple.browse                on                              default
zebbo/MACDATA-1/ZEBBO  com.apple.ignoreowner           on                              inherited from zebbo/MACDATA-1
zebbo/MACDATA-1/ZEBBO  com.apple.mimic                 off                             default
zebbo/MACDATA-1/ZEBBO  com.apple.devdisk               poolonly                        default
zebbo/MACDATA-1/ZEBBO  com.1bsnap:snapsync             true                            local
zebbo/MACDATA-1/ZEBBO  com.sun:auto-snapshot:monthly   false                           inherited from zebbo
zebbo/MACDATA-1/ZEBBO  com.sun:auto-snapshot:hourly    false                           inherited from zebbo
zebbo/MACDATA-1/ZEBBO  com.sun:auto-snapshot:weekly    true                            inherited from zebbo
zebbo/MACDATA-1/ZEBBO  com.sun:auto-snapshot:daily     false                           inherited from zebbo
zebbo/MACDATA-1/ZEBBO  com.sun:auto-snapshot:frequent  false                           inherited from zebbo
```

**Comment by lundman** [comment]
I assume you've seen:
https://openzfsonosx.org/wiki/FAQ#Q.29_Updating_crypto_to_2.0

It is an unfortunate side-effect that we got crypto before userobj quotas, and Linux the other way around.

**Comment by greg-fischer** [comment]
Oh wow! Thank you!
Simple enabling of 2 properties (like the FAQ says)
```
sudo zpool set feature@userobj_accounting=enabled pool
sudo zpool set feature@project_quota=enabled pool
```

Sorry.  I did a bit of googling and didn't find anything helpful.  All I went off of was "input/output error", ZFS, and Big Sur, which wasn't helpful.  Funny thing is, once I enabled userobj_accounting only, tried to mount, I got a very helpful message about enabling the properties required, because I didn't have project_quota.

Thank you!
